### ISSUE TITLE
DESKTOP_STARTUP_ID causes zathura, running inside tabbed, launched by urxvt, spawned by awesome, to fail rendering

### ISSUE THREAD

**Issue by dset0x** [awesomeWM/awesome#2385]
**Output of `awesome --version`:**
```
awesome v4.2 (Human after all)
 • Compiled against Lua 5.1.5 (running with Lua 5.1)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.5
 • LGI version: 0.9.0
```

**How to reproduce the issue:**
* spawn urxvt via awesome
*
```
read xid < <(tabbed -d)
zathura -e "$xid" 1.pdf &
zathura -e "$xid" 2.pdf
```

**Actual result:**

Nothing is displayed inside the (at least latter) zathura window.

**Why are you opening a bug here?**
My first guess was that this is a problem with zathura. I opened a [bug report](https://bugs.pwmt.org/project/zathura/issue/750) but they say they it shouldn't be executed with `DESKTOP_STARTUP_ID` set, according to the [freedesktop specification](https://specifications.freedesktop.org/startup-notification-spec/startup-notification-0.1.txt). Awesome sets this as described [here](https://awesomewm.org/doc/api/libraries/awful.spawn.html) (*The startup notification protocol* section).

Then I brought this up in the urxvt mailing list (detailed reply [here](http://lists.schmorp.de/pipermail/rxvt-unicode/2018q3/002473.html)). As I understand, (indeed) the spec doesn't really specify what failure is allowed and that urxvt has nothing to do with handling this variable. I cannot tell if awesome's behavior is really "abusing startup notifications" or if the problem is somewhere else in handling the new window.

I don't have a very good understanding of startup notifications or X, so apologies if I'm still in the wrong direction. I've worked around this as described in the bug reports, but it would be nice to see it fixed.

**Comment by psychon** [comment]
> I don't have a very good understanding of startup notifications or X, so apologies if I'm still in the wrong direction.

So, SN allows programs to say "I am beginning to start", "I am still starting", and "I am done starting, this is my window". Awesome sets `$DESKTOP_STARTUP_ID` through `libstartup-notification` (call to `sn_launcher_context_setup_child_process` in `spawn_callback` in `spawn.c`) and not directly by itself.

Since this has nothing to do with actually rendering windows, and since there is no feedback from awesome to the starting app, I do not really see what awesome could be doing differently here. Awesome only sets `$DESKTOP_STARTUP_ID` and doesn't do anything else that influences zathura. So, what other option is there besides just not enabling startup notification, which you can already do in your `rc.lua`?

Yet, I'm curious why Zathura would not render its window and will take a look.

**Comment by psychon** [comment]
Uhm. Could you try running `zathura` with (additionally) `GDK_DEBUG=all`? Sometimes (as in, about one in five attempts), this makes things work for me. The difference between the working and non-working attempts seems to be in Gtk seeing a ConfigureNotify and MapNotify event only sometimes, even though, according to `xwininfo`, the window is mapped in both cases.
Sadly, I didn't manage to capture a working version with `xtrace`, but in the non-working version that I got, Gdk clearly first creates the window and only changes the window's event mask to `value-list={event-mask=Exposure,VisibilityChange,StructureNotify,PropertyChange}` later. Thus, there is a race window here: If tabbed manages the window before Gtk changes its event mask, Gtk never sees that tabbed actually does its job.
Thus, I *guess* that this might actually be a bug in Gtk and instead Gtk should create the to-be-embedded window "elsewhere" and only reparent it into tabbed later (especially: after it set up its event mask).

**Comment by psychon** [comment]
So, this patch to GTK+ makes it so that problem *always* occurs for me:
```diff
diff --git a/gdk/x11/gdkwindow-x11.c b/gdk/x11/gdkwindow-x11.c
index 70daed9..d01d932 100644
--- a/gdk/x11/gdkwindow-x11.c
+++ b/gdk/x11/gdkwindow-x11.c
@@ -1113,6 +1113,9 @@ _gdk_x11_display_create_window_impl (GdkDisplay    *display,
                              0, window->depth, class, xvisual,
                              xattributes_mask, &xattributes);
 
+  XSync(xdisplay, False);
+  sleep(1);
+
   g_object_ref (window);
   _gdk_x11_display_add_window (x11_screen->display, &impl->xid, window);
 
```
And with this patch, the problem *never* occurs for me:
```diff
diff --git a/gdk/x11/gdkwindow-x11.c b/gdk/x11/gdkwindow-x11.c
index 70daed9..23eeed5 100644
--- a/gdk/x11/gdkwindow-x11.c
+++ b/gdk/x11/gdkwindow-x11.c
@@ -1048,6 +1048,9 @@ _gdk_x11_display_create_window_impl (GdkDisplay    *display,
   else
     xattributes.override_redirect = False;
 
+  xattributes.event_mask = StructureNotifyMask | PropertyChangeMask;
+  xattributes_mask |= CWEventMask;
+
   impl->override_redirect = xattributes.override_redirect;
 
   /* Sanity checks */
```
The root of the bug is that GTK creates its client window directly inside the embedder. Thus, there is a race between GTK setting up its window and at the same time the embedder starting to work with the window.

This precise bug happened when `tabbed` managed to map the window before GTK asked the X11 server "please tell me when my window is mapped". Thus, GTK never learnt about this happening.

My first patch guarantees this problem occurs by waiting one second between GTK creating the window and GTK asking for events on the window.
My second patch fixes this problem by immediately asking for events when the window is created. This might fix this particular race, but I bet there are still lots of other races possible. GTK really should create its window outside of the embedder and then reparent it when done.

> so apologies if I'm still in the wrong direction

No apology needed. I hope I gave you enough material to write a bug report to GTK. Please, leave a link to the GTK bug report here so that I (and everyone else) can follow it.

**Comment by psychon** [comment]
Actually... this is also (at least partly) a bug in `tabbed`. It ignores the `_XEMBED_INFO` property of its children. According to the xembed specification:

> [...] the client window must have a property called _XEMBED_INFO on it. 

The window created by GTK does not have this property set yet. The protocol does not really say what should happen in this case, but I would expect this to mean that `tabbed` should not touch the child window until it has an `_XEMBED_INFO` property.

I'll leave it between the authors of GTK and `tabbed` to decide who is at fault here (and my preferred solution would be for both of them to fix things).

**Comment by psychon** [comment]
Uhm... ping? Did you do any further actions?

**Comment by dset0x** [comment]
Hi @psychon. First of all, thank you _so much_ for doing so much work to figure out the cause of the issue, even though it is notabug as far as awesome is concerned. Really awesome of you. I think this is great spirit to work with.

I apologize for the long wait - work and studies took over all my time.

I just opened a bug [here](https://gitlab.gnome.org/GNOME/gtk/issues/1358) with some information and linked back to this thread.

**Comment by dset0x** [comment]
And here is a link to the suckless dev list - https://lists.suckless.org/dev/1809/32921.html

**Comment by psychon** [comment]
> I apologize for the long wait - work and studies took over all my time.

No problem. My intention was to wait some days and then open a bug report myself if I had to. Well, I didn't have the time to do just that and yesterday I stumbled upon this report again. A ping is done quickly, so I did just that. ;-)

Thanks for sticking with this bug despite everyone telling you "go away; report this elsewhere".

I'll eventually have some time to write some answer to the GTK bug that will hopefully get some attention. Right now, I guess they read "X11" and "GtkPlug" and defer the bug to the end of some todo list, because Wayland.
